File: services/categoryService.ts
import apiClient from './api';

export interface Category {
  id: number;
  user_id: number;
  name: string;
  type: 'income' | 'expense';
  icon?: string;
  color?: string;
  is_default: boolean;
}

export interface CreateCategoryRequest {
  name: string;
  type: 'income' | 'expense';
  icon?: string;
  color?: string;
}

export interface UpdateCategoryRequest {
  name?: string;
  type?: 'income' | 'expense';
  icon?: string;
  color?: string;
}

export interface CategoryResponse {
  data: Category[];
  total: number;
}

class CategoryService {
  async getCategories(): Promise<Category[]> {
    try {
      const response = await apiClient.get<{ success: boolean; data: Category[] }>('/categories');
      return response.data.data || [];
    } catch (error) {
      console.warn('Get categories error (route may not exist yet):', error);
      return [];
    }
  }

  async getCategoriesByType(type: 'income' | 'expense'): Promise<Category[]> {
    try {
      const response = await apiClient.get<CategoryResponse>('/categories', {
        params: { type },
      });
      return response.data.data;
    } catch (error) {
      console.error('Get categories by type error:', error);
      throw error;
    }
  }

  async getCategory(id: number): Promise<Category> {
    try {
      const response = await apiClient.get<{ data: Category }>(`/categories/${id}`);
      return response.data.data;
    } catch (error) {
      console.error('Get category error:', error);
      throw error;
    }
  }

  async createCategory(data: CreateCategoryRequest): Promise<Category> {
    try {
      const response = await apiClient.post<{ data: Category }>('/categories', data);
      return response.data.data;
    } catch (error) {
      console.error('Create category error:', error);
      throw error;
    }
  }

  async updateCategory(id: number, data: UpdateCategoryRequest): Promise<Category> {
    try {
      const response = await apiClient.put<{ data: Category }>(`/categories/${id}`, data);
      return response.data.data;
    } catch (error) {
      console.error('Update category error:', error);
      throw error;
    }
  }

  async deleteCategory(id: number): Promise<void> {
    try {
      await apiClient.delete(`/categories/${id}`);
    } catch (error) {
      console.error('Delete category error:', error);
      throw error;
    }
  }
}

export default new CategoryService();

File: backend/services/category_service.go
package services

import (
	"financial-freedom/models"
	"financial-freedom/repositories"
)

type CategoryService interface {
	CreateCategory(category *models.Category) error
	GetCategoryByID(id uint) (*models.Category, error)
	GetUserCategories(userID uint) ([]models.Category, error)
	UpdateCategory(category *models.Category) error
	DeleteCategory(id uint) error
	CreateDefaultCategories(userID uint) error
}

type categoryService struct {
	categoryRepository repositories.CategoryRepository
}

func NewCategoryService(categoryRepository repositories.CategoryRepository) CategoryService {
	return &categoryService{categoryRepository}
}

func (s *categoryService) CreateCategory(category *models.Category) error {
	return s.categoryRepository.Create(category)
}

func (s *categoryService) GetCategoryByID(id uint) (*models.Category, error) {
	return s.categoryRepository.FindByID(id)
}

func (s *categoryService) GetUserCategories(userID uint) ([]models.Category, error) {
	return s.categoryRepository.FindByUserID(userID)
}

func (s *categoryService) UpdateCategory(category *models.Category) error {
	return s.categoryRepository.Update(category)
}

func (s *categoryService) DeleteCategory(id uint) error {
	return s.categoryRepository.Delete(id)
}

// CreateDefaultCategories creates default categories for a new user
func (s *categoryService) CreateDefaultCategories(userID uint) error {
	defaultCategories := []models.Category{
		// Income categories
		{UserID: userID, Name: "GAJI", Type: "income", Icon: "icon-salary", Color: "#F0FDF4", IsDefault: true},
		{UserID: userID, Name: "INVESTASI", Type: "income", Icon: "icon-investment", Color: "#F0FDF4", IsDefault: true},
		{UserID: userID, Name: "BISNIS", Type: "income", Icon: "icon-business", Color: "#FFF7ED", IsDefault: true},

		// Expense categories
		{UserID: userID, Name: "MAKANAN", Type: "expense", Icon: "icon-food", Color: "#FEF3C7", IsDefault: true},
		{UserID: userID, Name: "TRANSPORTASI", Type: "expense", Icon: "icon-transport", Color: "#D1FAE5", IsDefault: true},
		{UserID: userID, Name: "BELANJA", Type: "expense", Icon: "icon-shopping", Color: "#FEE2E2", IsDefault: true},
		{UserID: userID, Name: "TAGIHAN", Type: "expense", Icon: "icon-bill", Color: "#E0E7FF", IsDefault: true},
		{UserID: userID, Name: "HIBURAN", Type: "expense", Icon: "icon-entertainment", Color: "#FCE7F3", IsDefault: true},
		{UserID: userID, Name: "KESEHATAN", Type: "expense", Icon: "icon-health", Color: "#ECFDF5", IsDefault: true},
		{UserID: userID, Name: "PENDIDIKAN", Type: "expense", Icon: "icon-education", Color: "#EFF6FF", IsDefault: true},
	}

	for _, category := range defaultCategories {
		if err := s.categoryRepository.Create(&category); err != nil {
			return err
		}
	}

	return nil
}
